src/cargo/core/compiler/custom_build.rs

```diff
@@ -679,47 +683,120 @@ impl BuildOutput {
+        const RESERVED_PREFIXES: &[&str] = &[
+            "rustc-flags=",
+            "rustc-link-lib=",
+            "rustc-link-search=",
+            "rustc-link-arg-cdylib=",
+            "rustc-link-arg-bins=",
+            "rustc-link-arg-bin=",
+            "rustc-link-arg-tests=",
+            "rustc-link-arg-benches=",
+            "rustc-link-arg-examples=",
+            "rustc-link-arg=",
+            "rustc-cfg=",
+            "rustc-check-cfg=",
+            "rustc-env=",
+            "warning=",
+            "rerun-if-changed=",
+            "rerun-if-env-changed=",
+        ];
+        const DOCS_LINK_SUGGESTION: &str = "See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                for more information about build script outputs.";
+
+        fn parse_directive<'a>(
+            whence: &str,
+            line: &str,
+            data: &'a str,
+            new_syntax: bool,
+            has_metadata_prefix: bool,
+        ) -> CargoResult<(&'a str, &'a str)> {
+            let mut iter = data.splitn(2, "=");
+            let key = iter.next();
+            let value = iter.next();
+            match (key, value) {
+                (Some(a), Some(b)) => Ok((a, b.trim_end())),
+                _ => bail!(
+                    "invalid output in {}: `{}`\n\
+                    Expected a line with `{}{}KEY=VALUE` with an `=` character, \
+                    but none was found.\n\
+                    {}",
+                    whence,
+                    line,
+                    if new_syntax { "cargo::" } else { "cargo:" },
+                    if has_metadata_prefix { "metadata=" } else { "" },
+                    DOCS_LINK_SUGGESTION
+                ),
+            }
+        }
+
         for line in input.split(|b| *b == b'\n') {
             let line = match str::from_utf8(line) {
                 Ok(line) => line.trim(),
                 Err(..) => continue,
             };
-            let mut iter = line.splitn(2, ':');
-            if iter.next() != Some("cargo") {
-                // skip this line since it doesn't start with "cargo:"
-                continue;
+            let mut new_syntax = false;
+            let (key, value) = if let Some(data) = line.strip_prefix("cargo::") {
+                new_syntax = true;
+
+                // For instance, `cargo::rustc-flags=foo`.
+                if RESERVED_PREFIXES
+                    .iter()
+                    .any(|prefix| data.starts_with(prefix))
+                {
+                    parse_directive(whence.as_str(), line, data, new_syntax, false)?
+                }
+                // For instance, `cargo::metadata=foo=bar`.
+                else if let Some(data) = data.strip_prefix("metadata=") {
+                    // Reserved keys should not be used with the `cargo::metadata=` syntax.
+                    if let Some(prefix) = RESERVED_PREFIXES
```

Comment:
Why do we need this error check?

Reply:
For example: `cargo::metadata=rustc-flags=xxx`. The `rustc-flags` is a reserved key. So I guess we don't want it to work with both `cargo::rustc-flags=` and `cargo::metadata::rustc-flags=` syntaxes.